### PR TITLE
Adds ability to sync with remote controller.

### PR DESCRIPTION
**Pull request by chriscox**
Allows syncing with a firebase-hosted remote controller of your choice.  This is the first of two PR's that will enable this functionality (see #2 below).

1) Client app will need to provide their firebase credentials as follows.
```javascript
// Example to forward your firebase config credentials to remixer.
remixer.inializeRemote({
    apiKey: "XXX",
    authDomain: "XXX.firebaseapp.com",
    databaseURL: "https://XXX.firebaseio.com"
});
```

2) Client user will need to click the share button to trigger the remote (to be available in a forthcoming PR).

### CODE REVIEW

File: src/core/Remixer.ts [hunk 1]
```diff
@@ -250,6 +251,9 @@ class Remixer {
         this.saveVariable(variable);
         variable.executeCallbacks();
       }
+
+      // Save remotely without throttling.
+      Remote.saveVariable(variable, false);
```
Comment (chuga) [inline src/core/Remixer.ts:257]: So what happens if I didn't set up or initialize the remote controller?

File: src/lib/Remote.ts [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+   */
+  private static _sharedInstance = new Remote();
+
+  /**
+   * Throttles network save calls.
+   * @private
+   * @static
+   * @type {any}
+   */
+  private static _throttle: any;
```
Comment (chuga) [inline src/lib/Remote.ts]: _throttledSaveVariable

File: src/lib/Remote.ts [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+   */
+  private dbReference(): firebase.database.Reference {
+    return firebase.database().ref(`remixer/${this.remoteId}`);
+  }
+
+  /**
+   * Start sharing the variable updates to the remote controller.
+   * @static
+   */
+  static startSharing(): void {
```
Comment (chuga) [inline src/lib/Remote.ts:132]: I thought we were going to start sharing automatically as soon as you enable it the first time? that's probably another preference we need to save

File: src/lib/Remote.ts [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    this._throttle = throttle(this._save, 300);
```
Comment (chuga) [inline src/lib/Remote.ts]: can you const this number at the top of the file?

File: src/lib/Remote.ts [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    return uuid().substring(0, 8);
+  }
+
+  /**
+   * Returns a database reference to the remixer instance.
+   * @private
+   * @return {firebase.database.Reference} The firebase database reference.
+   */
+  private dbReference(): firebase.database.Reference {
+    return firebase.database().ref(`remixer/${this.remoteId}`);
```
Comment (chuga) [inline src/lib/Remote.ts]: We should probably const the 'remixer/' string

**Comment by chuga** [comment]
Also, how do you feel about maybe squashing these commits into one?

File: src/core/Remixer.ts (hunk 1 above)
Comment (chriscox) [reply src/core/Remixer.ts:257]: In the remote class it checks if enabled, and if not  its a no-op. I Updated the comment to reflect this.

File: src/lib/Remote.ts (hunk 5 above)
Comment (chriscox) [reply src/lib/Remote.ts]: Done.

File: src/lib/Remote.ts (hunk 4 above)
Comment (chriscox) [reply src/lib/Remote.ts]: Done.

File: src/lib/Remote.ts (hunk 3 above)
Comment (chriscox) [reply src/lib/Remote.ts:132]: As we discussed offline, a call to this method will come from the share menu button (in an upcoming PR). We will then persist locally the sharing property for subsequent app refreshes.

**Review by chriscox — COMMENTED** [review]
Updated. PTAL.

**Review by chuga — APPROVED** [review]
lgtm

**Pull request merged**